src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -955,79 +1016,106 @@ void MergeTreeData::loadDataParts(bool skip_sanity_checks)
-                tryLogCurrentException(__PRETTY_FUNCTION__);
-            }
+        part->modification_time = part_disk_ptr->getLastModified(fs::path(relative_data_path) / part_name).epochTime();
+        /// Assume that all parts are Committed, covered parts will be detected and marked as Outdated later
+        part->setState(DataPartState::Committed);
 
-            /// Ignore broken parts that can appear as a result of hard server restart.
-            if (broken)
-            {
-                LOG_ERROR(log, "Detaching broken part {}{}. If it happened after update, it is likely because of backward incompability. You need to resolve this manually", getFullPathOnDisk(part_disk_ptr), part_name);
-                std::lock_guard loading_lock(mutex);
-                broken_parts_to_detach.push_back(part);
-                ++suspicious_broken_parts;
+        std::lock_guard loading_lock(mutex);
+        if (!data_parts_indexes.insert(part).second)
+            throw Exception("Part " + part->name + " already exists", ErrorCodes::DUPLICATE_DATA_PART);
 
-                return;
-            }
-            if (!part->index_granularity_info.is_adaptive)
-                has_non_adaptive_parts.store(true, std::memory_order_relaxed);
-            else
-                has_adaptive_parts.store(true, std::memory_order_relaxed);
+        addPartContributionToDataVolume(part);
+    };
 
-            part->modification_time = part_disk_ptr->getLastModified(fs::path(relative_data_path) / part_name).epochTime();
-            /// Assume that all parts are Committed, covered parts will be detected and marked as Outdated later
-            part->setState(DataPartState::Committed);
+    std::mutex part_select_mutex;
+    for (size_t thread = 0; thread < num_threads; ++thread)
+    {
+        pool.scheduleOrThrowOnError([&, thread]
+        {
+            while (true)
+            {
+                std::pair<String, DiskPtr> thread_part;
+                {
+                    const std::lock_guard lock{part_select_mutex};
 
-            std::lock_guard loading_lock(mutex);
-            if (!data_parts_indexes.insert(part).second)
-                throw Exception("Part " + part->name + " already exists", ErrorCodes::DUPLICATE_DATA_PART);
+                    if (remaining_thread_parts.empty())
+                        return;
 
-            addPartContributionToDataVolume(part);
+                    /// Steal task if nothing to do
+                    auto thread_idx = thread;
+                    if (threads_parts[thread].empty())
+                    {
+                        // Try random steal tasks from the next thread
+                        std::uniform_int_distribution<size_t> distribution(0, remaining_thread_parts.size() - 1);
+                        auto it = remaining_thread_parts.begin();
+                        std::advance(it, distribution(thread_local_rng));
+                        thread_idx = *it;
+                    }
```

Comment:
Doesn't this produce concurrent access? (due to index change, i.e. multiple threads can have the same index and will concurrently access to the `threads_parts`?)

Reply:
There is a https://github.com/amosbird/ClickHouse/blob/amos/src/Storages/MergeTree/MergeTreeData.cpp#L1096

Reply:
Oh, now I see `part_select_mutex`, got it, thanks!